This is a cherry-pick of https://github.com/triton-lang/triton/pull/7719 which
should make it into the next integration so this patch can simply be deleted
next integrate.

diff --git a/lib/Dialect/TritonGPU/Transforms/HoistTMEMAlloc.cpp b/lib/Dialect/TritonGPU/Transforms/HoistTMEMAlloc.cpp
--- a/lib/Dialect/TritonGPU/Transforms/HoistTMEMAlloc.cpp
+++ b/lib/Dialect/TritonGPU/Transforms/HoistTMEMAlloc.cpp
@@ -181,6 +181,11 @@ public:
       return failure();
     if (alloc->getBlock() != store->getBlock())
       return failure();
+    if (auto srcDef = store.getSrc().getDefiningOp()) {
+      if (alloc->getBlock() == srcDef->getBlock() &&
+          alloc->isBeforeInBlock(srcDef))
+        return failure();
+    }
     alloc.getSrcMutable().assign(store.getSrc());
     rewriter.replaceOp(store, alloc.getToken());
     return success();
